rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -0,0 +1,336 @@
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class RainbowCommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = router_address
+
+    def _resolve_asset(self, asset: Asset) -> EvmToken | CryptoAsset:
+        """Resolve a asset into evmtoken or cryptoasset
+
+        Args:
+            asset (Asset): Asset to be resolve
+
+        Returns:
+            EvmToken | CryptoAsset: Asset resolved
+        """
+        return self.evm_inquirer.native_token if asset == A_ETH else asset.resolve_to_evm_token()
+
+    def _tansform_decoded_events(
+            self,
+            decoded_events: list['EvmEvent'],
+            sender: ChecksumEvmAddress
+    ) -> tuple[Optional['EvmEvent'], Optional['EvmEvent']]:
+        """Transform the receive and spend events to rainbow events
+
+        Returns:
+            in_event (EvmEvent | None): input event, the receive the tokens
+            out_event (EvmEvent | None): ouput event, the spend the tokens
+        """
+        in_event = out_event = None
+        for event in decoded_events:
+            if (event.event_type == HistoryEventType.INFORMATIONAL and event.event_subtype == HistoryEventSubType.APPROVE):
+                # Remove the approve from rainbow router to aggretor
+                # to avoid show unnecessary events 
+                decoded_events.remove(event)
+
+            # Find the receive event
+            if (event.event_type == HistoryEventType.RECEIVE and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = CPT_RAINBOW_SWAPS
+                event.notes = f'Receive {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} as the result of a Rainbow Swap'  # noqa: E501
+                event.address = self.router_address
+                in_event = event
+
+            # Find the spend event
+            if (event.event_type == HistoryEventType.SPEND and event.event_subtype == HistoryEventSubType.NONE and event.location_label == sender):  # noqa: E501
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} in Rainbow Swap'  # noqa: E501
+                event.counterparty = CPT_RAINBOW_SWAPS
+                event.address = self.router_address
+                out_event = event
+
```

Comment:
what if any of them is None? You should log if something is off

Reply:
not every transaction will have the receiving event in decode_events, so in_event can be none, sometimes this event is added at the end of the process, that's why you can see that in some functions I create an actionitem to delete and create a new event for "receive"